sql/catalyst/src/main/java/org/apache/spark/sql/connector/metric/CustomFileTaskMetric.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.metric;
+
+import org.apache.spark.annotation.Evolving;
+
+/**
+ * A custom file task metric. This allows file based data source V2 implementations
+ * to use a single PartitionReader with multiple file readers. Each file reader can
+ * provide its own metrics values and they can be added into the parent PartitionReader
+ *
+ * @since 4.0.0
+ */
+@Evolving
+public interface CustomFileTaskMetric extends CustomTaskMetric {
```

Comment:
Hmm, I am wondering why we need this interface?

Reply:
I wasn't sure if it is OK to change an existing public interface. If it is acceptable to simply add this to `CustomTaskMetric` that would be perfect.

Reply:
I mean it seems feasible to use `CustomTaskMetric` to achieve same goal.

Reply:
+1 for @viirya 's advice.

Reply:
DSV1 uses `SQLMetric` which has a `set(v: Long)`  and an `add(v: Long)` so a data source can set an initial value for a metric and then keep updating it. DSV2 uses `CustomTaskMetric` which really should have an equivalent but doesn't. That's what this interface is trying to provide. 

In this case, `FilePartitionReader` needs to update the value of a `CustomTaskMetric` every time a file reader completes. So `FilePartitionReader` can have a metric say `ReadTime` which it updates every time a file reader completes. But, only the Parquet file reader provides this metric, so now `FilePartitionReader` can query the file reader and gets the metrics that are supported. But it has no way to update them. `FilePartitionReader` can of course create its own set of metrics corresponding to the supported ones, keep its own set of values and update them as needed, but having an interface to update the value which the metric supports seems so much more elegant.

Here's the usage of this interface - https://github.com/parthchandra/spark/blob/SPARK-47291/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FilePartitionReader.scala

And the corresponding metrics - 
https://github.com/parthchandra/spark/blob/SPARK-47291/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetMetricsV2.scala

If you still think this is not reasonable, I will try to re-work it without this interface.

Reply:
Actually, let me rework it without this interface